File: 1559-detect-cycles-in-2d-grid/1559-detect-cycles-in-2d-grid.py
class Solution:
    def containsCycle(self, grid: List[List[str]]) -> bool:
        visited = [[False for _ in range(len(grid[0]))] for _ in range(len(grid))]
        def dfs(i: int , j: int, prev: str, dir: int) -> int:
            if i < 0 or i == len(grid) or j < 0 or j == len(grid[0]) or grid[i][j] != prev:
                return False
            if visited[i][j]:
                return True
            else:
                visited[i][j] = True
                if dir != 1 and dfs(i + 1, j, grid[i][j], 0):
                    return True
                if dir != 0 and dfs(i - 1, j, grid[i][j], 1):
                    return True
                if dir != 3 and dfs(i, j + 1, grid[i][j], 2):
                    return True
                if dir != 2 and dfs(i, j - 1, grid[i][j], 3):
                    return True
            return False
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if not visited[i][j] and dfs(i, j, grid[i][j], -1):
                    return True
        return False

File: 1559-detect-cycles-in-2d-grid/1559-detect-cycles-in-2d-grid.cpp
class Solution {
public:
    bool containsCycle(vector<vector<char>>& grid) {
        vector<vector<bool>> visited(grid.size(), vector<bool>(grid[0].size(), false));
        for(int i = 0; i < grid.size(); i++)
        {
            for(int j = 0; j < grid[0].size(); j++)
            {
                if(!visited[i][j] && dfs(grid, visited, i, j, grid[i][j], -1))
                {
                    return true;
                }
            }
        }
        return false;
    }
private:
    bool dfs(vector<vector<char>>& grid, vector<vector<bool>>& visited, int i, int j, char prev, int dir)
    {
        if(i < 0 || i == grid.size() || j < 0 || j == grid[0].size() || grid[i][j] != prev)
        {
            return false;
        }
        if(visited[i][j])
        {
            return true;
        }
        else
        {
            visited[i][j] = true;
            if(dir != 1 && dfs(grid, visited, i + 1, j, grid[i][j], 0))
            {
                return true;
            }
            if(dir != 0 && dfs(grid, visited, i - 1, j, grid[i][j], 1))
            {
                return true;
            }
            if(dir != 3 && dfs(grid, visited, i, j + 1, grid[i][j], 2))
            {
                return true;
            }
            if(dir != 2 && dfs(grid, visited, i, j - 1, grid[i][j], 3))
            {
                return true;
            }
        }
        return false;
    }
};